File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let fastPointer = null
    let slowPointer = null
    
    // initialize fast and slow pointers
    slowPointer = head
    fastPointer = head
    for(let i=0; i<n; i++) {
        fastPointer = fastPointer.next
    }
    if(fastPointer !== null) {
        while(fastPointer.next !== null) {
            fastPointer = fastPointer.next
            slowPointer = slowPointer.next
        }
    
        // remove the node next to slow pointer
        slowPointer.next = slowPointer.next.next   
    } else {
        head = head.next
    }
    return head
    
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode slow = head;
        ListNode fast = head;
        ListNode prev = null;
        
        for(int i=0; i<n; i++) {
            if(fast != null)
                fast = fast.next;
        }
        while(fast != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next;
        }

        if(prev != null)
            prev.next = slow.next;
        else {
            // prev = slow.next;
            // if prev is null, slow has to be at the 1st pos
            head = head.next;
        }
        return head;
    }
}